File: source_code/guppy.cpp
#include "guppy.hpp"
#include "oop.hpp"
#include <cmath>
#include <random>
#include "linkedList.hpp"


guppy::guppy() : fish(){
// ctor default
}


guppy::guppy (int x,int y) : fish(x,y) {
// ctor user-defined
}


guppy::guppy (const guppy& g):fish(g) {
// cctor
}

guppy& guppy::operator=(const guppy& other){
    growth = other.growth;
    fullRate = other.fullRate;
    orientation = other.orientation;
    foodEaten = other.foodEaten;
    pos = other.pos;
    return *this;
}


guppy::~guppy(){
// dtor
}

void guppy::move(double diff,linkedList<food>& listFood){

  if (isStarving() && !listFood.isEmpty()){
      // Jika lapar, guppy bergerak mendekati makanan (jika ada)
	    findFood(diff,listFood); 
  } else {	
     // bergerak bebas
	   fish::move(diff);
  }

  fullRate-=diff; // tingkat kekenyangan berkurang

  if (cos(orientation) > 0) {
    if (growth == 1) {
      draw_image("ikanLv1Right.png", pos.first, pos.second);
    } else if (growth == 2) {
      draw_image("ikanLv2RightC.png", pos.first, pos.second);
    } else {
      draw_image("ikanLv3RightC.png", pos.first, pos.second);
    }
  } else {
    if (growth == 1) {
      draw_image("ikanLv1Left.png", pos.first, pos.second);
    } else if (growth == 2) {
      draw_image("ikanLv2LeftC.png", pos.first, pos.second);
    } else {
      draw_image("ikanLv3LeftC.png", pos.first, pos.second);
    }
  }
}


void guppy::findFood(double diff , linkedList<food>& listFood){
// Saat lapar, Guppy akan mendekati makanan ikan yang ada di akuarium
	food Near = this->getNearestFood(listFood);
	orientation = atan2(Near.getOrdinat() - this->getOrdinat() , Near.getAbsis() - this->getAbsis()) * 180/3.14159265;
	
  pos.first += SPEED*diff*cos(orientation);
  pos.second += SPEED*diff*sin(orientation);

  eat(Near, listFood);
}


void guppy::eat(food& feed,linkedList<food>& listFood){
// Guppy memakan makanan ikan yang berada pada radius makannya
  if (!listFood.isEmpty()){
    if (this->getDistance(feed)<=FOOD_RADIUS){
      fullRate=FULL_TIME;
      foodEaten++;
      if (foodEaten==FOOD_NEEDED && growth<3){
        growth++;
        foodEaten=0;
      }
      listFood.remove(feed);
    }
  }
}


void guppy::produceCoin(double diff,linkedList<coin>& listCoin){
// Guppy mengeluarkan koin setiap periode tertentu
  if (time >= COIN_TIME){
		coin produced(growth*5,pos.first,pos.second);
		time = 0;
		listCoin.add(produced);
	} else{
		time+= diff;
	}
}

food guppy::getNearestFood(linkedList<food>& feed) {
// Mengembalikan food terdekat ke guppy pada listFood
    node<food>* temp = feed.getHead();
    food minFood = temp->info;
    temp = temp->next;
    while (temp != NULL) {
      if (this->getDistance(temp->info) < this->getDistance(minFood)) {
        minFood = temp->info;
      }
      temp = temp->next;
    }
    return minFood;
}

File: Java DOG/guppy.java
public class Guppy extends Fish{
    private double time;
    private static final int COIN_TIME=7;
    public Guppy(){
        time=0;
    }
    public Guppy(int x,int y){
        super(x,y);
        time=0;
    }
    public Boolean equals(Guppy other){
        return super.equals(other) && (this.time==other.time); 
    }
    public void move(double diff,LinkedList<Food> listFood){
        if (isStarving() && !listFood.isEmpty()){
            // Jika lapar, Guppy bergerak mendekati makanan (jika ada)
            findFood(diff,listFood); 
        } else {	
           // bergerak bebas
            super.move(diff);
        }
        fullRate-=diff; // tingkat kekenyangan berkurang
        if (Math.cos(orientation) > 0) {
            if (growth == 1) {
                // draw_image("ikanLv1Right.png", pos.first, pos.second);
            } else if (growth == 2) {
                // draw_image("ikanLv2RightC.png", pos.first, pos.second);
            } else {
                // draw_image("ikanLv3RightC.png", pos.first, pos.second);
            }
        } else {
            if (growth == 1) {
                // draw_image("ikanLv1Left.png", pos.first, pos.second);
            } else if (growth == 2) {
                // draw_image("ikanLv2LeftC.png", pos.first, pos.second);
            } else {
                // draw_image("ikanLv3LeftC.png", pos.first, pos.second);
            }
        }
    }
    public void findFood(double diff , LinkedList<Food> listFood){
        // Saat lapar, Guppy akan mendekati makanan ikan yang ada di akuarium
        Food Near = this.getNearestFood(listFood);
        orientation = Math.atan2(Near.getOrdinat() - this.getOrdinat() , Near.getAbsis() - this.getAbsis()) * 180/3.14159265;
        setPos(this.getAbsis() + SPEED*diff*Math.cos(orientation), this.getOrdinat() + SPEED*diff*Math.sin(orientation));
        eat(Near, listFood);
    }
    public void eat(Food feed,LinkedList<Food> listFood){
        // Guppy memakan makanan ikan yang berada pada radius makannya
        if (!listFood.isEmpty()){
            if (this.getDistance(feed)<=FOOD_RADIUS){
                fullRate=FULL_TIME;
                foodEaten++;
                if (foodEaten==FOOD_NEEDED && growth<3){
                    growth++;
                    foodEaten=0;
                }
                listFood.remove(feed);
            }
        }
    }
    public void produceCoin(double diff,LinkedList<Coin> listCoin){
        // Guppy mengeluarkan koin setiap periode tertentu
        if (time >= COIN_TIME){
            Coin produced = new Coin(growth*5,this.getAbsis(),this.getOrdinat());
            time = 0;
            listCoin.add(produced);
        } else{
            time+= diff;
        }
    }
    public Food getNearestFood(LinkedList<Food> listFood) {
        // Mengembalikan Food terdekat ke Guppy pada listFood
        Node<Food> temp = listFood.getHead();
        Food minFood = temp.getData();
        temp = temp.getNext();
        while (temp != null) {
            if (this.getDistance(temp.getData()) < this.getDistance(minFood)) {
                minFood = temp.getData();
            }
            temp = temp.getNext();
        }
        return minFood;
    }

    public static void main(String [] args){
        Guppy g = new Guppy();
        System.out.println(g.toString());
        LinkedList<Food> listFood = new LinkedList<Food>();
        g.move(0.03,listFood);
        System.out.println(g.toString());
    }
}